content/articles/2019-08-callr-task-q.Rmd

```diff
@@ -0,0 +1,586 @@
+---
+title: Multi Process Task Queue in 100 Lines of R Code
+slug: callr-task-q
+author: Gábor Csárdi
+date: '2019-08-03'
+description: >
+  We use `callr::r_session` to implement a worker pool and task queue in
+  100 lines of R code.
+categories:
+- programming
+- parallelism
+photo:
+  url: https://unsplash.com/photos/cDwZ40Lj9eo
+  author: Davide Ragusa
+---
+
+```{r setup, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE, comment = "#>"
+)
+```
+
+```{r, comment="", results="asis", echo = FALSE}
+if (knitr::opts_knit$get("rmarkdown.pandoc.to") == "html") {
+  old.hooks <- fansi::set_knit_hooks(knitr::knit_hooks)
+  options(crayon.enabled = TRUE)
+  options(crayon.colors = 256)
+  invisible(crayon::num_colors(forget = TRUE))
+}
+```
+
+## Introduction
+
+```{r echo = FALSE}
+source("figure/taskq.R")
+```
+
+This post is a demo of `callr::r_session`, a persistent R session you can
+use to run R code asynchronously. I set out to build a task queue, which
+runs tasks in subprocesses, concurrently, in mere 100 lines of R code.
+
+Here is a short teaser for how the queue will work. `task_q$new()` creates
+a new R6 object, which represents the queue. Its `push()` method adds a
+task, which is a function and its arguments, similar to `callr::r()`.
+The `pop()` method gets the results of the first task that has finished.
+`pop()` has a timeout argument, which lets you wait for a task to finish,
+if all pushed tasks are still running. It returns `NULL` if no task has
+finished before the timeout was over. The timeout can be `0` or `Inf`,
+meaning no wait at all, or wait indefinitely. The default timeout is `0`.
+
+```{r}
+q <- task_q$new()
+q$push(function() { Sys.getpid() })
+q$push(function() { Sys.sleep(.5); Sys.getpid() })
+q$pop()
+```
+
+This `pop()` call returned `NULL`, as none of the tasks are done yet.
+Even though running `Sys.getpid()` is fast, the worker processes also need
+200-500 ms time start up, when the queue is created. If you are willing to
+wait a bit, at least one task should be done in less than half a second,
+but usually not the second one yet:
+
+```{r}
+sec <- as.difftime(1, units = "secs")
+q$pop(sec * 1/2)$result
```

Comment:
Can you not just supply an integer here? I think the use of difftime is a bit of a distraction